lib/spack/spack/test/cmd/env.py

```diff
@@ -1255,6 +1301,108 @@ def test_env_without_view_install(tmpdir, mock_stage, mock_fetch, install_mocker
     check_mpileaks_and_deps_in_view(view_dir)
 
 
+def setup_combined_multiple_env():
+    env("create", "test1")
+    test1 = ev.read("test1")
+    with test1:
+        add("mpileaks")
+    test1.concretize()
+
+    env("create", "test2")
+    test2 = ev.read("test2")
+    with test2:
+        add("libelf")
+    test2.concretize()
+
+    env("create", "--include-concrete", "test1", "--include-concrete", "test2", "combined_env")
+    combined = ev.read("combined_env")
+
+    return test1, test2, combined
+
+
+def test_env_include_concrete_env_yaml():
+    env("create", "test")
+    test = ev.read("test")
+
+    with test:
+        add("mpileaks")
+
+    env("create", "--include-concrete", "test", "combined_env")
+
+    combined = ev.read("combined_env")
+    combined_yaml = ev.config_dict(combined.raw_yaml)
+
+    assert "include_concrete" in combined_yaml
+    assert test.path in combined_yaml["include_concrete"]
+
+
+def test_env_include_concrete_env_path_yaml():
+    env("create", "test")
+    test = ev.read("test")
+
+    with test:
+        add("mpileaks")
+
+    env("create", "--include-concrete", test.path, "combined_env")
+
+    combined = ev.read("combined_env")
+    combined_yaml = ev.config_dict(combined.raw_yaml)
+
+    assert "include_concrete" in combined_yaml
+    assert test.path in combined_yaml["include_concrete"]
+
+
+def test_env_include_bad_concrete_env():
+    with pytest.raises(ev.SpackEnvironmentError):
+        env("create", "--include-concrete", "nonexistant_env", "combined_env")
+
+
+def test_env_include_multiple_concrete_envs():
+    test1, test2, combined = setup_combined_multiple_env()
+
+    combined_yaml = ev.config_dict(combined.raw_yaml)
+
+    assert test1.path in combined_yaml["include_concrete"][0]
+    assert test2.path in combined_yaml["include_concrete"][1]
+
+    # No local specs in the combined env
+    assert not combined_yaml["specs"]
+
+
+def test_env_include_concrete_envs_lockfile():
+    test1, test2, combined = setup_combined_multiple_env()
+
+    with open(combined.lock_path) as f:
+        lockfile_as_dict = combined._read_lockfile(f)
+
+    assert lockfile_as_dict["include"][test1.path]["roots"][0]["hash"] in test1.specs_by_hash
+    assert lockfile_as_dict["include"][test2.path]["roots"][0]["hash"] in test2.specs_by_hash
```

Comment:
Could we make these checks a little stricter by comparing 

```suggestion
set(entry["hash"] for entry in lockfile_as_dict["include"][test1.path]["roots"]) == set(test1.specs_by_hash)
set(entry["hash"] for entry in lockfile_as_dict["include"][test2.path]["roots"]) == set(test2.specs_by_hash)